README.md
# eredis

Non-blocking Redis client with focus on performance and robustness.

[![Build Status](https://github.com/Nordix/eredis/workflows/CI/badge.svg)](https://github.com/Nordix/eredis)
[![Hex pm](https://img.shields.io/hexpm/v/eredis.svg?style=flat)](https://hex.pm/packages/eredis)
[![Hex.pm](https://img.shields.io/hexpm/dt/eredis.svg)](https://hex.pm/packages/eredis)

This fork is the official continuation of `wooga/eredis`.
It includes several improvements, such as
TLS support, Sentinel support and TCP error handling corrections.
See [CHANGELOG.md](CHANGELOG.md) for details.

Note: This client connects to a standalone Redis node, optionally using Sentinel.
For *Redis Cluster*, you need
[eredis_cluster](https://github.com/Nordix/eredis_cluster) or
[ecredis](https://github.com/HalloAppInc/ecredis).

Supported Redis features:

 * Any command, through `eredis:q/2,3`
 * Transactions
 * Pipelining
 * Authentication & multiple DBs
 * Pubsub
 * TLS
 * Sentinel support

Generated API documentation: [doc/eredis.md](doc/eredis.md)

Published documentation can also be found on [hexdocs](https://hexdocs.pm/eredis/).

## Setup

If you have Redis running on localhost with default settings, like:

    docker run --rm --net=host redis:latest

you may copy and paste the following into a shell to try out Eredis:

    git clone https://github.com/Nordix/eredis.git
    cd eredis
    rebar3 shell
    {ok, C} = eredis:start_link().
    {ok, <<"OK">>} = eredis:q(C, ["SET", "foo", "bar"]).
    {ok, <<"bar">>} = eredis:q(C, ["GET", "foo"]).

To connect to a Redis instance listening on a Unix domain socket:

```erlang
{ok, C1} = eredis:start_link({local, "/var/run/redis.sock"}, 0).
```

To connect to a Redis instance using TLS:

```erlang
Options = [{tls, [{cacertfile, "ca.crt"},
                  {certfile,   "client.crt"},
                  {keyfile,    "client.key"}]}],
{ok, C2} = eredis:start_link("127.0.0.1", ?TLS_PORT, Options),
```

To connect to a Redis sentinel cluster:

```erlang
SentinelOptions = [{master_group, mymaster},
                   {endpoints, [{"127.0.0.1", 26379}]}],
{ok, C1} = eredis:start_link([{sentinel, SentinelOptions}]).
```

## Example

MSET and MGET:

```erlang
KeyValuePairs = ["key1", "value1", "key2", "value2", "key3", "value3"].
{ok, <<"OK">>} = eredis:q(C, ["MSET" | KeyValuePairs]).
{ok, Values} = eredis:q(C, ["MGET" | ["key1", "key2", "key3"]]).
```

HASH

```erlang
HashObj = ["id", "objectId", "message", "message", "receiver", "receiver", "status", "read"].
{ok, <<"OK">>} = eredis:q(C, ["HMSET", "key" | HashObj]).
{ok, Values} = eredis:q(C, ["HGETALL", "key"]).
```

LIST

```erlang
eredis:q(C, ["LPUSH", "keylist", "value"]).
eredis:q(C, ["RPUSH", "keylist", "value"]).
eredis:q(C, ["LRANGE", "keylist", 0, -1]).
```

Transactions:

```erlang
{ok, <<"OK">>} = eredis:q(C, ["MULTI"]).
{ok, <<"QUEUED">>} = eredis:q(C, ["SET", "foo", "bar"]).
{ok, <<"QUEUED">>} = eredis:q(C, ["SET", "bar", "baz"]).
{ok, [<<"OK">>, <<"OK">>]} = eredis:q(C, ["EXEC"]).
```

Pipelining:

```erlang
P1 = [["SET", a, "1"],
      ["LPUSH", b, "3"],
      ["LPUSH", b, "2"]].
[{ok, <<"OK">>}, {ok, <<"1">>}, {ok, <<"2">>}] = eredis:qp(C, P1).
```

Pubsub:

```erlang
1> eredis_sub:sub_example().
received {subscribed,<<"foo">>,<0.34.0>}
{<0.34.0>,<0.37.0>}
2> eredis_sub:pub_example().
received {message,<<"foo">>,<<"bar">>,<0.34.0>}
ok
3>
```

Pattern Subscribe:

```erlang
1> eredis_sub:psub_example().
received {subscribed,<<"foo*">>,<0.33.0>}
{<0.33.0>,<0.36.0>}
2> eredis_sub:ppub_example().
received {pmessage,<<"foo*">>,<<"foo123">>,<<"bar">>,<0.33.0>}
ok
3>
```

## Commands

### Query: [qp/2,3](doc/eredis.md#q-2)

Eredis has one main function to interact with redis, which is
`eredis:q(Client::pid(), Command::iolist())`. The response will either
be `{ok, Value::binary() | [binary()]}` or `{error,
Message::binary()}`.  The value is always the exact value returned by
Redis, without any type conversion. If Redis returns a list of values,
this list is returned in the exact same order without any type
conversion.

### Pipelined query: [qp/2,3](doc/eredis.md#qp-2)

To send multiple requests to redis in a batch, aka. pipelining
requests, you may use `eredis:qp(Client::pid(),
[Command::iolist()])`. This function returns `{ok, [Value::binary()]}`
where the values are the redis responses in the same order as the
commands you provided.

### Connect a client: [start_link/1](doc/eredis.md#start_link-1)

To start the client, use `start_link/1` or one of its variants. `start_link/1`
takes the following options (proplist):

* `host`: DNS name or IP address as string; or unix domain socket as `{local,
  Path}` (available in OTP 19+)
* `port`: integer, default is 6379
* `database`: integer or 0 for default database, default: 0
* `username`: string, default: no username
* `password`: string, default: no password
* `reconnect_sleep`: integer of milliseconds to sleep between reconnect attempts, default: 100
* `connect_timeout`: timeout value in milliseconds to use in the connect, default: 5000
* `socket_options`: proplist of [gen_tcp](https://erlang.org/doc/man/gen_tcp.html)
  options used when connecting the socket, default is `?SOCKET_OPTS`.
  It's possible to specify `{active, N}` or `{active, true}`.
  Active `once` and `false` are not supported. The default is N = 10.
* `tls`: enable TLS by providing a list of
  [options](https://erlang.org/doc/man/ssl.html) used when establishing the TLS
  connection, default is off.
  It's possible to specify `{active, N}` or `{active, true}` in TLS options.
  Active `once` and `false` are not supported. The default is N = 10.
  Note that `{active, N}` with TLS requires OTP 21.3 or later.

## Implicit pipelining

Commands are pipelined automatically so multiple processes can share the same
Eredis connection instance. Although `q/2,3` and `qp/2,3` are blocking until the
response is returned, Eredis is not blocked.

```
  Process A          Process B          Eredis        TCP/TLS socket
     |                  |                  |          (Redis server)
     | q(Pid, Command1) |                  |                 |
     |------------------------------------>|---------------->|
     |                  | q(Pid, Command2) |                 |
     |                  |----------------->|---------------->|
     |                  |                  |                 |
    ...                ...                ...               ...
     |                  |                  |                 |
     |                  |                  |      Response 1 |
     |<------------------------------------|<----------------|
     |                  |                  |      Response 2 |
     |                  |<-----------------|<----------------|
```

## Reconnecting on Redis down / network failure / timeout / etc

When Eredis for some reason loses the connection to Redis, Eredis
will keep trying to reconnect until a connection is successfully
established, which includes the `AUTH` and `SELECT` calls. The sleep
time between attempts to reconnect can be set in the
`eredis:start_link/1` call.

As long as the connection is down, Eredis will respond to any request
immediately with `{error, no_connection}` without actually trying to
connect. This serves as a kind of circuit breaker and prevents a
stampede of clients just waiting for a failed connection attempt or
`gen_server:call` timeout.

Note: If Eredis is starting up and cannot connect, it will fail
immediately with `{connection_error, Reason}`.

## Pubsub

Thanks to Dave Peticolas (jdavisp3), eredis supports
pubsub. [`eredis_sub`](doc/eredis_sub.md) offers a separate client that will forward
channel messages from Redis to an Erlang process in an "active-once"
pattern similar to gen_tcp sockets. After every message received, the
controlling process must acknowledge receipt using
`eredis_sub:ack_message/1`.

If the controlling process does not process messages fast enough,
eredis will queue the messages up to a certain queue size controlled
by configuration. When the max size is reached, eredis will either
drop messages or crash, also based on configuration.

Subscriptions are managed using `eredis_sub:subscribe/2` and
`eredis_sub:unsubscribe/2`. When Redis acknowledges the change in
subscription, a message is sent to the controlling process for each
channel. Then, eredis sends a message on the form `{subscribed, Channel,
ClientPid}` to the controlling process, which must be acked using
`eredis_sub:ack_message/1`.

eredis also supports Pattern Subscribe using `eredis_sub:psubscribe/2`
and `eredis_sub:punsubscribe/2`. As with normal subscriptions, a message
is sent to the controlling process for each channel and eredis sends a
message on the form `{subscribed, Channel, ClientPid}` to the
controlling process, which must be acked using `eredis_sub:ack_message/1`.

The controlling process is also notified in case of reconnection attempts or
failures. See `test/eredis_pubsub_SUITE.erl` for examples.

Starting with version 1.5, eredis automatically resubscribes after reconnect.
Then, the controlling process will receive `{subscribed, Channel, ClientPid}`
messages again for every channel and pattern subscribed to. These must also be
acked.

Here is a list of all the messages that are sent to the controlling process.
Some messages don't need to be acked, but it does not harm doing so.

* `{eredis_connected, Pid}` when the socket to Redis is established
  and authenticated. Doesn't need to be acked.

* `{eredis_disconnected, Pid}` when the connection to Redis has been lost.
  Doesn't need to be acked.

* `{eredis_reconnect_attempt, Pid}` at every reconnect attempt, when the
  connection to Redis has been lost. Doesn't need to be acked.

* `{eredis_reconnect_failed, Pid, {error, {connection_error, Reason}}}` after
  every failed reconnect attempt. Doesn't need to be acked.

* `{message, Channel, Message, Pid}` for every incoming message on subscribed
  channels. Needs to be acked using `eredis_sub:ack_message/1`.

* `{pmessage, Pattern, Channel, Message, Pid}` for every incoming message on
  channels subscribed to by pattern, using using `eredis_sub:psubscribe/2`.
  Needs to be acked using `eredis_sub:ack_message/1`.

* `{subscribed, Channel, Pid}` when a subscription has been confirmed by Redis.
  Channel is either a channel (subscribe) or a pattern (psubscribe). Needs to be
  acked using `eredis_sub:ack_message/1`.

* `{unsubscribed, Channel, Pid}` when a subscription has been removed from
  Redis. Channel is either a channel (unsubscribe) or a pattern (punsubscribe).
  Needs to be acked using `eredis_sub:ack_message/1`.

## AUTH and SELECT

Eredis also implements the AUTH and SELECT calls for you. When the
client is started with something else than default values for password
and database, it will issue the `AUTH` and `SELECT` commands
appropriately, even when reconnecting after a timeout.

## Benchmarking

Using [lasp-bench](https://github.com/lasp-lang/lasp-bench/) you may
benchmark Eredis on your own hardware using the provided config and
driver. See `priv/basho_bench_driver_eredis.config` and
`src/basho_bench_driver_eredis.erl`.

Testcase summary from our daily runs:

* [eredis](https://bjosv.github.io/eredis-benchmark/results/latest/eredis/summary.png)
* [eredis_pipeline](https://bjosv.github.io/eredis-benchmark/results/latest/eredis_pipeline/summary.png)

The [eredis-benchmark](https://github.com/bjosv/eredis-benchmark) repo runs
a daily job that produces above graphs. It also contains the script
`run-tests.sh` that might help you with the needed steps when setting up the
benchmark testing on your own.

## Queueing

Eredis uses the same queueing mechanism as Erldis. `eredis:q/2` uses
`gen_server:call/2` to do a blocking call to the client
gen_server. The client will immediately send the request to Redis, add
the caller to the queue and reply with `noreply`. This frees the
gen_server up to accept new requests and parse responses as they come
on the socket.

When data is received on the socket, we call `eredis_parser:parse/2`

CHANGELOG.md
# CHANGELOG

## v1.7.0

July 2022.

* Replaces `error_logger` with `logger`.

* Turns repeated reconnection errors into log level `notice`.

* Parser optimizations.

* Enables performance tuning of the received packets handling.
  The socket is set to `{active, N}` with N = 10 by default.
  This is configurable by including `{active, N}` in the socket options or
  in the TLS options when TLS is used.

  Note that `{active, N}` with TLS requires OTP 21.3 or later.
  When using OTP below 21.3 the option needs to be set to `{active, true}`.

## v1.6.0

July 2022.

* Adds sentinel support.

* Obfuscates username and password in the state, to prevent them from appearing
  verbatim in logs and stack traces. They can be provided as 0-ary functions
  returning the actual secret when applied.

* Fix a crash that happens when reconnect fails twice with two different
  reasons. This bug was introduced in v1.5.0.

## v1.5.1

May 2022.

* Solves a bug introduced in v1.5.0 causing the reconnect process to survive
  if the client is stopped while attempting to reconnect.

* Drops the spawned process used for reconnecting. Instead, the reconnect
  attempts are scheduled using timers.

  The option `reconnect_sleep` now applies to the time between a
  successful connect and the first reconnect attemt, if the connection is lost
  just after connecting. However, there is no delay before reconnecting
  if the connection has been up for at least `reconnect_sleep` milliseconds.

## v1.5.0

May 2022.

* No delay before the first reconnect attempt.

* eredis_sub: Automatic re-subscribe on reconnect. Messages on the
  form `{subscribed, Channel, Pid}` are sent to the controlling
  process in this case and need to be acked.

* eredis_sub: TLS, custom TCP options, AUTH with username, SELECT
  database and registered name options added.

## v1.4.1

June 2021.

* Restore support for rebar 2.
* Optimize parser.

## v1.4.0

June 2021.

* Support for named connection processes.
* Support AUTH with username.

## v1.3.3

March 2021.

* Include correct files in published package on hex.pm.

## v1.3.2

February 2021.

* Official release

## v1.3.1-nordix

February 2021.

* Fix build problems with Mix and other non-Rebar tools

## v1.3.0-nordix

December 2020 in the Nordix fork by Björn Svensson and Viktor Söderqvist.

* TLS support
* Correction regarding chunked error responses
* All socket errors, including send and setopts errors, trigger reconnect
* Explicitly close socket before reconnect, preventing dangling ports
* Terminate reconnect loop when client terminates (fixes wooga/eredis#124)
* Try all addresses from `inet:getaddrs/2` instead of just one when connecting
* Stray messages don't crash the connection process
* Improved tests and coverage
* CI builds using Github Actions
* Elvis code style and Dialyzer corrections

## v1.2.0

2018 in the original `wooga/eredis` repo.

* Unix domain socket support (PR wooga/eredis#108 by Igor Slepchin)
* Reset parser state on disconnect (PR wooga/eredis#106 by Ivan Baidakou)
* Non-blocking init AKA async connect (PR wooga/eredis#105 by Valery Meleshkin)
* Some improved tests

## v1.1.0

* Merged a ton of of old and neglected pull requests. Thanks to
  patient contributors:
  * Emil Falk
  * Evgeny Khramtsov
  * Kevin Wilson
  * Luis Rascão
  * Аверьянов Илья (savonarola on github)
  * ololoru
  * Giacomo Olgeni

* Removed rebar binary, made everything a bit more rebar3 & mix
  friendly.


## v1.0.8

* Fixed include directive to work with rebar 2.5.1. Thanks to Feng Hao
  for the patch.

## v1.0.7

* If an eredis_sub_client needs to reconnect to Redis, the controlling
  process is now notified with the message `{eredis_reconnect_attempt,
  Pid}`. If the reconnection attempt fails, the message is
  `{eredis_reconnect_failed, Pid, Reason}`. Thanks to Piotr Nosek for
  the patch.

* No more deprecation warnings of the `queue` type on OTP 17. Thanks
  to Daniel Kempkens for the patch.

* Various spec fixes. Thanks to Hernan Rivas Acosta and Anton Kalyaev.

## v1.0.6

* If the connection to Redis is lost, requests in progress will
  receive `{error, tcp_closed}` instead of the `gen_server:call`
  timing out. Thanks to Seth Falcon for the patch.

## v1.0.5

* Added support for not selecting any specific database. Thanks to
  Mikl Kurkov for the patch.

## v1.0.4

* Added `eredis:q_noreply/2` which sends a fire-and-forget request to
  Redis. Thanks to Ransom Richardson for the patch.

* Various type annotation improvements, typo fixes and robustness
  improvements. Thanks to Michael Gregson, Matthew Conway and Ransom
  Richardson.

## v1.0.3

* Fixed bug in eredis_sub where when the connection to Redis was lost,
  the socket would not be set into {active, once} on reconnect. Thanks
  to georgeye for the patch.

## v1.0.2

* Fixed bug in eredis_sub where the socket was incorrectly set to
  `{active, once}` twice. At large volumes of messages, this resulted
  in too many messages from the socket and we would be unable to keep
  up. Thanks to pmembrey for reporting.

## v1.0

* Support added for pubsub thanks to Dave Peticolas
  (jdavisp3). Implemented in `eredis_sub` and `eredis_sub_client` is a
  subscriber that will forward messages from Redis to an Erlang
  process with flow control. The user can configure to either drop
  messages or crash the driver if a certain queue size inside the
  driver is reached.

* Fixed error handling when eredis starts up and Redis is still
  loading the dataset into memory.

## v0.7.0

* Support added for pipelining requests, which allows batching
  multiple requests in a single call to eredis. Thanks to Dave
  Peticolas (jdavisp3) for the implementation.

## v0.6.0

* Support added for transactions, by Dave Peticolas (jdavisp3) who implemented
  parsing of nested multibulks.

## v0.5.0

* Configurable reconnect sleep time, by Valentino Volonghi (dialtone)

* Support for using eredis as a poolboy worker, by Valentino Volonghi
  (dialtone)


src/eredis_sub_client.erl
%%
%% eredis_pubsub_client
%%
%% This client implements a subscriber to a Redis pubsub channel. It
%% is implemented in the same way as eredis_client, except channel
%% messages are streamed to the controlling process. Messages are
%% queued and delivered when the client acknowledges receipt.
%%
%% There is one consuming process per eredis_sub_client.
%% @private
-module(eredis_sub_client).
-behaviour(gen_server).
-include("eredis.hrl").
-include("eredis_sub.hrl").
-define(CONNECT_TIMEOUT, 5000).
-define(RECONNECT_SLEEP, 100).

%% API
-export([start_link/1, stop/1]).

%% gen_server callbacks
-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
         terminate/2, code_change/3]).

%%
%% API
%%

-spec start_link(eredis_sub:sub_options()) ->
          {ok, Pid::pid()} | {error, Reason::term()}.
start_link(Options) ->
    case proplists:lookup(name, Options) of
        {name, Name} ->
            gen_server:start_link(Name, ?MODULE, Options, []);
        none ->
            gen_server:start_link(?MODULE, Options, [])
    end.

stop(Pid) ->
    gen_server:call(Pid, stop).

%%====================================================================
%% gen_server callbacks
%%====================================================================

init(Options) ->
    %% Options
    Host           = proplists:get_value(host, Options, "127.0.0.1"),
    Port           = proplists:get_value(port, Options, 6379),
    Database       = proplists:get_value(database, Options, 0),
    Username       = proplists:get_value(username, Options, undefined),
    Password       = proplists:get_value(password, Options, undefined),
    ReconnectSleep = proplists:get_value(reconnect_sleep, Options, ?RECONNECT_SLEEP),
    ConnectTimeout = proplists:get_value(connect_timeout, Options, ?CONNECT_TIMEOUT),
    SocketOptions  = proplists:get_value(socket_options, Options, []),
    TlsOptions     = proplists:get_value(tls, Options, []),
    Transport      = case TlsOptions of
                         [] -> gen_tcp;
                         _ -> ssl
                     end,
    Active         = eredis_client:get_active(SocketOptions, TlsOptions),

    %% eredis_pub specific options
    MaxQueueSize   = proplists:get_value(max_queue_size, Options, infinity),
    QueueBehaviour = proplists:get_value(queue_behaviour, Options, drop),

    State = #state{host            = Host,
                   port            = Port,
                   database        = eredis_client:read_database(Database),
                   auth_cmd        = eredis_client:get_auth_command(Username,
                                                                    Password),
                   reconnect_sleep = ReconnectSleep,
                   connect_timeout = ConnectTimeout,
                   socket_options  = SocketOptions,
                   tls_options     = TlsOptions,
                   transport       = Transport,
                   active          = Active,
                   channels        = [],
                   pchannels       = [],
                   parser_state    = eredis_parser:init(),
                   msg_queue       = queue:new(),
                   max_queue_size  = MaxQueueSize,
                   queue_behaviour = QueueBehaviour},

    %% Same sync/async connect behaviour as eredis_client.
    case ReconnectSleep of
        no_reconnect ->
            case connect(State) of
                {ok, _NewState} = Res -> Res;
                {error, Reason} -> {stop, Reason}
            end;
        T when is_integer(T) ->
            self() ! initiate_connection,
            {ok, State}
    end.

%% Set the controlling process. All messages on all channels are directed here.
handle_call({controlling_process, Pid}, _From, State) ->
    case State#state.controlling_process of
        undefined ->
            ok;
        {OldRef, _OldPid} ->
            erlang:demonitor(OldRef)
    end,
    Ref = erlang:monitor(process, Pid),
    {reply, ok, State#state{controlling_process={Ref, Pid}, msg_state = ready}};

handle_call(get_channels, _From, State) ->
    {reply, {ok, State#state.channels ++ State#state.pchannels}, State};


handle_call(stop, _From, State) ->
    {stop, normal, ok, State};

handle_call(_Request, _From, State) ->
    {reply, unknown_request, State}.


%% Controlling process acks, but we have no connection. When the
%% connection comes back up, we should be ready to forward a message
%% again.
handle_cast({ack_message, Pid},
            #state{controlling_process={_, Pid}, socket = undefined} = State) ->
    {noreply, State#state{msg_state = ready}};

%% Controlling process acknowledges receipt of previous message. Send
%% the next if there is any messages queued or ask for more on the
%% socket.
handle_cast({ack_message, Pid},
            #state{controlling_process={_, Pid}} = State) ->
    NewState = case queue:out(State#state.msg_queue) of
                   {empty, _Queue} ->
                       State#state{msg_state = ready};
                   {{value, Msg}, Queue} ->
                       send_to_controller(Msg, State),
                       State#state{msg_queue = Queue, msg_state = need_ack}
               end,
    {noreply, NewState};

handle_cast({subscribe, Pid, Channels},
            #state{transport = Transport,
                   socket = Socket,
                   controlling_process = {_, Pid}} = State) ->
    ok = send_subscribe_command(Transport, Socket, "SUBSCRIBE", Channels),
    NewChannels = add_channels(Channels, State#state.channels),
    {noreply, State#state{channels = NewChannels}};


handle_cast({psubscribe, Pid, Channels},
            #state{transport = Transport,
                   socket = Socket,
                   controlling_process = {_, Pid}} = State) ->
    ok = send_subscribe_command(Transport, Socket, "PSUBSCRIBE", Channels),
    NewChannels = add_channels(Channels, State#state.pchannels),
    {noreply, State#state{pchannels = NewChannels}};



handle_cast({unsubscribe, Pid, Channels},
            #state{transport = Transport,
                   controlling_process = {_, Pid}} = State) ->
    Command = eredis:create_multibulk(["UNSUBSCRIBE" | Channels]),
    ok = Transport:send(State#state.socket, Command),
    NewChannels = remove_channels(Channels, State#state.channels),
    {noreply, State#state{channels = NewChannels}};



handle_cast({punsubscribe, Pid, Channels},
            #state{transport = Transport,
                   controlling_process = {_, Pid}} = State) ->
    Command = eredis:create_multibulk(["PUNSUBSCRIBE" | Channels]),
    ok = Transport:send(State#state.socket, Command),
    NewChannels = remove_channels(Channels, State#state.pchannels),
    {noreply, State#state{pchannels = NewChannels}};



handle_cast({ack_message, _}, State) ->
    {noreply, State};

handle_cast(_Msg, State) ->
    {noreply, State}.


%% Receive TCP/TLS data from socket. Match `Socket' to enforce sanity.
handle_info({Type, Socket, Bs}, #state{socket = Socket} = State)
  when Type =:= tcp; Type =:= ssl->
    NewState = handle_response(Bs, State),
    case queue:len(NewState#state.msg_queue) > NewState#state.max_queue_size of
        true ->
            case State#state.queue_behaviour of
                drop ->
                    Msg = {dropped, queue:len(NewState#state.msg_queue)},
                    send_to_controller(Msg, NewState),
                    {noreply, NewState#state{msg_queue = queue:new()}};
                exit ->
                    {stop, max_queue_size, State}
            end;
        false ->
            {noreply, NewState}
    end;

%% Socket switched to passive mode due to {active, N}.
handle_info({Passive, Socket},
            #state{socket = Socket, transport = Transport, active = N} = State)
  when Passive =:= tcp_passive; Passive =:= ssl_passive ->
    case setopts(Socket, Transport, [{active, N}]) of
        ok ->
            {noreply, State};
        {error, Reason} ->
            maybe_reconnect(Reason, State)
    end;

handle_info({Error, Socket, _Reason}, #state{socket = Socket} = State)
  when Error =:= tcp_error; Error =:= ssl_error ->
    maybe_reconnect(Error, State);

%% Socket got closed, for example by Redis terminating idle
%% clients. If desired, spawn of a new process which will try to reconnect and
%% notify us when Redis is ready. In the meantime, we can respond with
%% an error message to all our clients.
handle_info({Closed, Socket}, #state{socket = OurSocket} = State)
  when Closed =:= tcp_closed orelse Closed =:= ssl_closed,
       Socket =:= OurSocket orelse Socket =:= fake_socket ->
    send_to_controller({eredis_disconnected, self()}, State),
    maybe_reconnect(Closed, State#state{socket = undefined});

handle_info(initiate_connection,
            #state{socket = undefined,
                   reconnect_sleep = ReconnectSleep} = State) ->
    case connect(State) of
        {ok, NewState} ->
            {noreply, NewState};
        {error, Reason} ->
            erlang:send_after(ReconnectSleep, self(), {reconnect, Reason}),
            {noreply, State}
    end;

handle_info({reconnect, Reason}, #state{socket = undefined} = State) ->
    %% Scheduled reconnect, if disconnected.
    maybe_reconnect(Reason, State#state{reconnect_timer = undefined});
handle_info({reconnect, _Reason}, State) ->
    %% Already connected.
    {noreply, State#state{reconnect_timer = undefined}};

%% Our controlling process is down.
handle_info({'DOWN', Ref, process, Pid, _Reason},
            #state{controlling_process={Ref, Pid}} = State) ->
    {stop, shutdown, State#state{controlling_process=undefined,
                                 msg_state=ready,
                                 msg_queue=queue:new()}};

%% eredis can be used in Poolboy, but it requires to support a simple API
%% that Poolboy uses to manage the connections.
handle_info(stop, State) ->
    {stop, shutdown, State};

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, #state{socket = undefined}) ->
    ok;
terminate(_Reason, #state{socket = Socket, transport = Transport}) ->
    Transport:close(Socket).

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%%--------------------------------------------------------------------
%%% Internal functions
%%--------------------------------------------------------------------

%% @doc When no channels are given, we unsubscribe from all channels. This
%% matches the semantics of (P)UNSUBSCRIBE without channels.
-spec remove_channels([binary()], [binary()]) -> [binary()].
remove_channels([], _OldChannels) ->
    [];
remove_channels(Channels, OldChannels) ->
    lists:foldl(fun lists:delete/2, OldChannels, Channels).

-spec add_channels([binary()], [binary()]) -> [binary()].
add_channels(Channels, OldChannels) ->
    lists:foldl(fun(C, Cs) ->
                        case lists:member(C, Cs) of
                            true ->
                                Cs;
                            false ->
                                [C|Cs]
                        end
                end, OldChannels, Channels).

%% @doc Sends a subscribe or psubscribe command to Redis.
-spec send_subscribe_command(Transport :: gen_tcp | ssl,
                             Socket :: gen_tcp:socket() | ssl:socket(),
                             Command :: iodata(),
                             Channels :: list()) -> ok.
send_subscribe_command(_Transport, _Socket, _Command, []) ->
    ok;
send_subscribe_command(Transport, Socket, Command, Channels) ->
    Cmd = eredis:create_multibulk([Command | Channels]),
    Transport:send(Socket, Cmd).

-spec handle_response(Data::binary(), State::#state{}) -> NewState::#state{}.
%% @doc: Handle the response coming from Redis. This should only be
%% channel messages that we should forward to the controlling process
%% or queue if the previous message has not been acked. If there are
%% more than a single response in the data we got, queue the responses
%% and serve them up when the controlling process is ready
handle_response(Data, #state{parser_state = ParserState} = State) ->
    case eredis_parser:parse(ParserState, Data) of
        {ReturnCode, Value, NewParserState} ->
            reply({ReturnCode, Value},
                  State#state{parser_state=NewParserState});

        {ReturnCode, Value, Rest, NewParserState} ->
            NewState = reply({ReturnCode, Value},
                             State#state{parser_state=NewParserState}),
            handle_response(Rest, NewState);

        {continue, NewParserState} ->
            State#state{parser_state = NewParserState}
    end.

%% @doc: Sends a reply to the controlling process if the process has
%% acknowledged the previous process, otherwise the message is queued
%% for later delivery.
reply({ok, [<<"message">>, Channel, Message]}, State) ->
    queue_or_send({message, Channel, Message, self()}, State);

reply({ok, [<<"pmessage">>, Pattern, Channel, Message]}, State) ->
    queue_or_send({pmessage, Pattern, Channel, Message, self()}, State);



reply({ok, [<<"subscribe">>, Channel, _]}, State) ->
    queue_or_send({subscribed, Channel, self()}, State);

reply({ok, [<<"psubscribe">>, Channel, _]}, State) ->
    queue_or_send({subscribed, Channel, self()}, State);


reply({ok, [<<"unsubscribe">>, Channel, _]}, State) ->
    queue_or_send({unsubscribed, Channel, self()}, State);


reply({ok, [<<"punsubscribe">>, Channel, _]}, State) ->
    queue_or_send({unsubscribed, Channel, self()}, State);
reply({ReturnCode, Value}, State) ->
    throw({unexpected_response_from_redis, ReturnCode, Value, State}).


queue_or_send(Msg, State) ->
    case State#state.msg_state of
        need_ack ->
            MsgQueue = queue:in(Msg, State#state.msg_queue),
            State#state{msg_queue = MsgQueue};
        ready ->
            send_to_controller(Msg, State),
            State#state{msg_state = need_ack}
    end.


%% @doc: Helper for connecting to Redis. These commands are
%% synchronous and if Redis returns something we don't expect, we
%% crash. Returns {ok, State} or {error, Reason}.
connect(#state{host = Host, port = Port, socket_options = SocketOptions,
               transport = Transport, active = Active,
               connect_timeout = ConnectTimeout, tls_options = TlsOptions,
               auth_cmd = AuthCmd, database = Db} = State) ->
    case eredis_client:connect(Host, Port, SocketOptions, TlsOptions,
                               ConnectTimeout, AuthCmd, Db, Active) of
        {ok, Socket} ->
            %% Re-subscribe to channels. Channels are stored in reverse order in
            %% state.
            ok = send_subscribe_command(Transport, Socket, "SUBSCRIBE",
                                        lists:reverse(State#state.channels)),
            ok = send_subscribe_command(Transport, Socket, "PSUBSCRIBE",
                                        lists:reverse(State#state.pchannels)),

            %% Notify application that connection is ready.
            send_to_controller({eredis_connected, self()}, State),

            %% In case the connection terminates immediately (this happens with
            %% an expired certificate with TLS 1.3) schedule a reconnect already
            %% so that we don't try to reconnect if an error is received before
            %% reconnect_sleep milliseconds has elapsed.
            {ok, schedule_reconnect(unknown, State#state{socket = Socket})};
        Error ->
            Error
    end.

close_socket(#state{socket = undefined} = State) ->
    State;
close_socket(#state{socket = Socket, transport = Transport} = State) ->
    send_to_controller({eredis_disconnected, self()}, State),
    Transport:close(Socket),
    %% Throw away the socket. The absence of a socket is used to
    %% signal we are "down"; discard possibly patrially parsed data
    State#state{socket = undefined, parser_state = eredis_parser:init()}.

%% @doc Schedules a reconnect attempt, if reconnect is enabled.
-spec schedule_reconnect(Reason :: any(), #state{}) -> #state{}.
schedule_reconnect(_Reason, #state{reconnect_sleep = no_reconnect} = State) ->
    State;
schedule_reconnect(Reason, #state{reconnect_sleep = ReconnectSleep,
                                  reconnect_timer = undefined} = State) ->
    TRef = erlang:send_after(ReconnectSleep, self(), {reconnect, Reason}),
    State#state{reconnect_timer = TRef}.

%% @doc Reconnects, but not if a reconnect has been scheduled or if reconnect is
%% disabled. The socket in the state is closed, if any. Returns {noreply, State}
%% or {stop, ExitReason, State} like handle_info.
maybe_reconnect(_Reason, #state{reconnect_sleep = no_reconnect} = State) ->
    %% If we aren't going to reconnect, then there is nothing else for this
    %% process to do.
    {stop, normal, close_socket(State)};
maybe_reconnect(_Reason, #state{reconnect_timer = TRef} = State)
  when is_reference(TRef) ->
    %% Reconnect already scheduled.
    {noreply, close_socket(State)};
maybe_reconnect(_Reason, State) ->
    State1 = close_socket(State),

    %% Controller might want to be notified about every reconnect attempt
    send_to_controller({eredis_reconnect_attempt, self()}, State1),

    case connect(State1) of
        {ok, State2} ->
            {noreply, State2};
        {error, Reason} ->
            %% Controller might want to be notified about every reconnect
            %% failure and reason
            send_to_controller({eredis_reconnect_failed, self(),
                                {error, {connection_error, Reason}}}, State1),
            {noreply, schedule_reconnect(Reason, State1)}
    end.

setopts(Socket, _Transport=gen_tcp, Opts) -> inet:setopts(Socket, Opts);
setopts(Socket, _Transport=ssl, Opts)     ->  ssl:setopts(Socket, Opts).

send_to_controller(_Msg, #state{controlling_process=undefined}) ->
    ok;
send_to_controller(Msg, #state{controlling_process={_Ref, Pid}}) ->
    Pid ! Msg.
